### PR TITLE
BUG: Overflow warning when using Series.diff() on Series with Periods

### LINKED ISSUE #58320: BUG:  Using `Series.diff()` on a Series with Periods on Windows shows an overflow error
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
>>> import pandas as pd
>>> pd.Series(
...                     [pd.Period("2012-1-1", freq="D"), pd.Period("2012-1-2", freq="D")]
...                 ).diff()
C:\Anaconda3\envs\pandasstubs\lib\site-packages\pandas\core\arrays\datetimelike.py:1306: RuntimeWarning: overflow encountered in scalar multiply
  new_data = np.array([self.freq.base * x for x in new_i8_data])
0      NaT
1    <Day>
dtype: object
```

### Issue Description

This error only appears on Windows, not on Linux!

Discovered in a `pandas-stubs` pull request:  https://github.com/pandas-dev/pandas-stubs/pull/907/files#r1571049325

### Expected Behavior

No error message

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit                : d9cdd2ee5a58015ef6f4d15c7226110c9aab8140
python                : 3.9.16.final.0
python-bits           : 64
OS                    : Windows
OS-release            : 10
Version               : 10.0.19045
machine               : AMD64
processor             : Intel64 Family 6 Model 158 Stepping 13, GenuineIntel
byteorder             : little
LC_ALL                : None
LANG                  : None
LOCALE                : English_United States.1252

pandas                : 2.2.2
numpy                 : 1.26.4
pytz                  : 2024.1
dateutil              : 2.9.0.post0
setuptools            : 68.2.2
pip                   : 22.3.1
Cython                : None
pytest                : 8.1.1
hypothesis            : None
sphinx                : None
blosc                 : None
feather               : None
xlsxwriter            : 3.2.0
lxml.etree            : 5.2.1
html5lib              : 1.1
pymysql               : None
psycopg2              : None
jinja2                : 3.1.3
IPython               : None
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : 4.12.3
bottleneck            : None
dataframe-api-compat  : None
fastparquet           : None
fsspec                : None
gcsfs                 : None
matplotlib            : 3.8.3
numba                 : None
numexpr               : 2.10.0
odfpy                 : None
openpyxl              : 3.1.2
pandas_gbq            : None
pyarrow               : 15.0.2
pyreadstat            : 1.2.7
python-calamine       : None
pyxlsb                : 1.0.10
s3fs                  : None
scipy                 : 1.13.0
sqlalchemy            : 2.0.29
tables                : 3.9.1
tabulate              : 0.9.0
xarray                : 2024.3.0
xlrd                  : 2.0.1
zstandard             : 0.22.0
tzdata                : 2024.1
qtpy                  : None
pyqt5                 : None
</details>

### PR DESCRIPTION
**Pull request by Aloqeely**
- [x] closes #58320 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: doc/source/whatsnew/v3.0.0.rst [hunk 1]
```diff
@@ -416,15 +416,15 @@ MultiIndex
...
+- Bug in :class:`DataFrame` and :class:`Series` ``repr`` of :py:class:`collections.abc.Mapping` elements. (:issue:`57915`)
 - Bug in :meth:`DataFrame.to_dict` raises unnecessary ``UserWarning`` when columns are not unique and ``orient='tight'``. (:issue:`58281`)
 - Bug in :meth:`DataFrame.to_excel` when writing empty :class:`DataFrame` with :class:`MultiIndex` on both axes (:issue:`57696`)
 - Bug in :meth:`DataFrame.to_string` that raised ``StopIteration`` with nested DataFrames. (:issue:`16098`)
 - Bug in :meth:`read_csv` raising ``TypeError`` when ``index_col`` is specified and ``na_values`` is a dict containing the key ``None``. (:issue:`57547`)
 
 Period
 ^^^^^^
--
+- Bug in :meth:`Series.diff` showing ``RuntimeWarning`` of overflowing when the Series has :class:`Period` (:issue:`58320`)
```
Comment (Aloqeely) [inline doc/source/whatsnew/v3.0.0.rst:427]: Should I mention this happens on Windows only?

File: pandas/tests/extension/test_period.py [hunk 2]
```diff
@@ -104,11 +102,7 @@ def check_reduce(self, ser: pd.Series, op_name: str, skipna: bool):
 
     @pytest.mark.parametrize("periods", [1, -2])
     def test_diff(self, data, periods):
```
Comment (mroeschke) [inline pandas/tests/extension/test_period.py:104]: I think this whole test function can be removed if its just calling super

File: pandas/core/arrays/datetimelike.py [hunk 3]
```diff
@@ -1303,16 +1303,18 @@ def _sub_periodlike(self, other: Period | PeriodArray) -> npt.NDArray[np.object_
         self._check_compatible_with(other)
 
         other_i8, o_mask = self._get_i8_values_and_mask(other)
-        new_i8_data = add_overflowsafe(self.asi8, np.asarray(-other_i8, dtype="i8"))
-        new_data = np.array([self.freq.base * x for x in new_i8_data])
-
+        new_data = add_overflowsafe(
+            self.asi8, np.asarray(-other_i8, dtype="i8")
+        ).astype(object)
```
Comment (mroeschke) [inline pandas/core/arrays/datetimelike.py:1308]: Not a great solution if this always needs to be cast to `object`

Comment (Aloqeely) [reply pandas/core/arrays/datetimelike.py:1308]: Why not? The comment says "we return an object-dtype ndarray" so it makes sense to cast it to `object`.
I thought this is better than creating a new array which will be object dtype (old behavior)

Comment (WillAyd) [reply pandas/core/arrays/datetimelike.py:1308]: I agree this seems suboptimal, and that comment seems pretty far detached from where the object conversion happens right? I agree this is a bit suspect; the old code seems to use primitive types so overall this is adding a good deal of overhead.

Do you know where the cast to object is actually occurring?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/datetimelike.py (hunk 3 above)
Comment (Aloqeely) [reply pandas/core/arrays/datetimelike.py:1308]:
```python
new_data = np.array([self.freq.base * x for x in new_i8_data])
```
Here, `self.freq.base` is what causes `new_data` to be object dtype

Comment (Aloqeely) [reply pandas/core/arrays/datetimelike.py:1308]: Running a quick check, it seems my solution has the same performance as the old code.
```python
In [1]: import pandas as pd
In [2]: pr = pd.period_range(start='2017-01-01', periods=100, freq="D")
In [3]: %timeit pd.Series(pr).diff()
280 µs ± 1.14 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```
Compared to old code:
```python
In [3]: %timeit pd.Series(pr).diff()
\pandas\core\arrays\datetimelike.py:1307: RuntimeWarning: overflow encountered in scalar multiply
  new_data = np.array([self.freq.base * x for x in new_i8_data])
357 µs ± 1.85 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Pull request closed without merging**